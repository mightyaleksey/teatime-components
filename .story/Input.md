## Preview


### size s

```
<div>
  some text for the baseline
  <Input
    name='input-s'
    size='s'
    style={{width: '200px'}}/>
  {' '}
  <Input
    disabled={true}
    name='input-s-d'
    size='s'
    style={{width: '200px'}}/>
</div>
```


### size m

```
<div>
  some text for the baseline
  <Input
    name='input-m'
    size='m'
    style={{width: '200px'}}/>
  {' '}
  <Input
    disabled={true}
    name='input-m-d'
    size='m'
    style={{width: '200px'}}/>
</div>
```


### size l

```
<div>
  some text for the baseline
  <Input
    name='input-l'
    placeholder='Input something...'
    size='l'
    style={{width: '200px'}}/>
  {' '}
  <Input
    defaultValue='Disabled text'
    disabled={true}
    name='input-l-d'
    size='l'
    style={{width: '200px'}}/>
</div>
```
