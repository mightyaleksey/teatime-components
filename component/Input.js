'use strict';

const {Component, PropTypes} = require('react');
const {filterProps, isControlled, genericName, nullToString} = require('../lib/util');
const {isUndefined, noop} = require('../lib/dash');
const React = require('react');
const cc = require('classnames');
const warn = require('../lib/warn');

const cssModules = {
  l: require('../style/input/input-l.css'),
  m: require('../style/input/input-m.css'),
  s: require('../style/input/input-s.css'),
};

let didWarnForInputDefaultValue = false;

class Input extends Component {
  constructor(props) {
    super(props);

    this._controlled = isControlled(props);

    if (process.env.NODE_ENV !== 'production')
      if (this._controlled && !isUndefined(props.defaultValue) && !didWarnForInputDefaultValue) {
        didWarnForInputDefaultValue = true;
        warn(true, 'defaultValue', 'Input', props.type);
      }

    const value = this._controlled
      ? props.value
      : props.defaultValue;

    this.state = {
      value: isUndefined(value) ? '' : nullToString(value),
    };
  }

  componentWillReceiveProps(nextProps) {
    this._controlled = isControlled(nextProps);

    this.setState({
      value: this._controlled
        ? nullToString(nextProps.value)
        : this.state.value,
    });
  }

  css = tokenName => genericName(this.props, tokenName)

  focus() {
    if (!this._input) return;
    this._input.focus();
  }

  select() {
    if (!this._input) return;
    this._input.select();
  }

  _onChange = e => {
    const value = e.target.value;

    if (!this._controlled) this.setState({value});
    this.props.onChange(e, {value});
  }

  _onClear = e => {
    const value = '';

    if (!this._controlled) this.setState({value});
    this.props.onChange(e, {value});
    this.focus();
  }

  render() {
    const {className} = this.props;
    const {css} = this;

    return (
      <span
        {...filterProps(this.props)}
        className={cc(css('wrapper'), className)}
        onChange={void 0}>
        {this.renderClear()}
        {this.renderInput()}
      </span>
    );
  }

  renderClear() {
    const {disabled} = this.props;
    const {value} = this.state;

    if (disabled || !value) return null;

    const {css} = this;

    return (
      <span
        className={css('clear')}
        onClick={this._onClear}/>
    );
  }

  renderInput() {
    const {
      autoComplete,
      autoFocus,
      disabled,
      id,
      maxLength,
      name,
      placeholder,
      readOnly,
      type,
    } = this.props;

    const {value} = this.state;
    const {css} = this;

    return (
      <input
        autoComplete={autoComplete}
        autoFocus={autoFocus}
        className={css('control')}
        disabled={disabled}
        id={id}
        maxLength={maxLength}
        name={name}
        onChange={this._onChange}
        placeholder={placeholder}
        readOnly={readOnly}
        ref={r => this._input = r}
        type={type}
        value={value}/>
    );
  }
}

Input.defaultProps = {
  onChange: noop,
  size: 'm',
  styles: cssModules,
  type: 'text',
};

Input.propTypes = {
  autoComplete: PropTypes.oneOf([
    'off',
    'on',
  ]),
  autoFocus: PropTypes.bool,
  className: PropTypes.string,
  defaultValue: PropTypes.string,
  disabled: PropTypes.bool,
  id: PropTypes.string,
  maxLength: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
  ]),
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  placeholder: PropTypes.string,
  readOnly: PropTypes.bool,
  size: PropTypes.oneOf([
    'l',
    'm',
    's',
  ]),
  styles: PropTypes.object,
  type: PropTypes.string,
  value: PropTypes.string,
};

module.exports = Input;
