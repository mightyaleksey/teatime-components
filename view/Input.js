'use strict';

const { Component, PropTypes } = require('react');
const { bind, hasValueProp, sanitizeProps } = require('../tool/component');
const { isUndefined, noop } = require('../tool/func');
const { style, styleName } = require('../tool/className');
const React = require('react');
const warning = require('../tool/warning');

var didWarnForDefaultValue = false;

class Input extends Component {
  constructor(props) {
    super(props);

    this.controlled = hasValueProp(props);

    if (process.env.NODE_ENV !== 'production' && this.controlled && !didWarnForDefaultValue) { // eslint-disable-line no-undef
      warning(isUndefined(props.defaultValue),
        'Input contains an input of type %s with both value and defaultValue props. ' +
        'Input elements must be either controlled or uncontrolled ' +
        '(specify either the value prop, or the defaultValue prop, but not ' +
        'both). Decide between using a controlled or uncontrolled input ' +
        'element and remove one of these props. More info: ' +
        'https://fb.me/react-controlled-components',
        props.type);

      didWarnForDefaultValue = true;
    }

    const value = this.controlled
      ? props.value
      : props.defaultValue;

    this.state = {
      value: !isUndefined(value)
        ? value
        : '',
    };

    bind(this, [
      'onChange',
      'onClearClick',
    ]);
  }

  componentWillReceiveProps(nextProps) {
    if (this.controlled) {
      this.setState({value: nextProps.value});
    }
  }

  focus() {
    if (this.refs.control) {
      this.refs.control.focus();
    }
  }

  onChange(e) {
    this.updateValue(e, e.target.value);
  }

  onClearClick(e) {
    if (this.props.disabled) {
      return;
    }

    this.updateValue(e, '');
    this.focus();
  }

  select() {
    if (this.refs.control) {
      this.refs.control.select();
    }
  }

  updateValue(e, value) {
    if (!this.controlled) {
      this.setState({value});
    }

    this.props.onChange(e, {value});
  }

  render() {
    return (
      <span className={styleName(this.props)}>
        {this.renderInput()}
        {this.renderClear()}
      </span>
    );
  }

  renderClear() {
    if (!this.state.value || this.props.disabled) {
      return null;
    }

    return (
      <span
        className={this.props.styles.clear}
        onClick={this.onClearClick}/>
    );
  }

  renderInput() {
    return (
      <input
        {...sanitizeProps(['styleName', 'styles'], ...this.props)}
        className={style(this.props.styles, 'control', {
          hasClear: this.state.value,
        })}
        defaultValue={undefined} // Cause we have a controlled input
        onChange={this.onChange}
        ref='control'
        value={this.state.value}/>
    );
  }
}

Input.defaultProps = {
  onChange: noop,
  styleName: 'wrapper',
  styles: {},
  type: 'text',
};

Input.propTypes = {
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  styleName: PropTypes.string,
  styles: PropTypes.shape({
    clear: PropTypes.string.isRequired,
    control: PropTypes.string.isRequired,
    wrapper: PropTypes.string,
  }),
  type: PropTypes.oneOf([
    'password',
    'text',
  ]),
};

module.exports = Input;
