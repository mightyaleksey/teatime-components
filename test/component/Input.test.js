import Input from '../../component/Input';
import React from 'react';
import renderer from 'react-test-renderer';

test('Input renders correctly with defaults', () => {
  const tree = renderer.create(
    <Input
      name='input control'/>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

test('Input renders correctly with defined props', () => {
  const tree = renderer.create(
    <Input
      autoComplete={false}
      autoFocus={true}
      className='outer'
      disabled={true}
      id='inputs id'
      maxLength={10}
      name='input control'
      placeholder='it is a placeholder'
      readOnly={false}
      type='password'
      value='custom value'/>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
