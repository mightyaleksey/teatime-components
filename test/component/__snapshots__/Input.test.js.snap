// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders correctly with defaults 1`] = `
<span
  className="wrapper"
  onChange={undefined}
>
  <input
    autoComplete={undefined}
    autoFocus={undefined}
    className="control"
    disabled={undefined}
    id={undefined}
    maxLength={undefined}
    name="input control"
    onChange={[Function]}
    placeholder={undefined}
    readOnly={undefined}
    type="text"
    value=""
  />
</span>
`;

exports[`Input renders correctly with defined props 1`] = `
<span
  className="wrapper outer"
  onChange={undefined}
>
  <input
    autoComplete="off"
    autoFocus={true}
    className="control"
    disabled={true}
    id="inputs id"
    maxLength={10}
    name="input control"
    onChange={[Function]}
    placeholder="it is a placeholder"
    readOnly={false}
    type="password"
    value="custom value"
  />
</span>
`;
